validator/client/beacon-api/doppelganger.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	// Retrieve current epoch
+	headers, err := c.getHeaders(ctx)
+	if err != nil || headers == nil || headers.Data == nil || len(headers.Data) == 0 ||
+		headers.Data[0].Header == nil || headers.Data[0].Header.Message == nil {
+		return nil, errors.Wrapf(err, "failed to get headers")
+	}
+
+	headSlotUint64, err := strconv.ParseUint(headers.Data[0].Header.Message.Slot, 10, 64)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse head slot")
+	}
+
+	headSlot := types.Slot(headSlotUint64)
+	currentEpoch := slots.ToEpoch(headSlot)
+
+	// Extract input pubkeys we did not validate for the 2 last epochs.
+	// If we detect onchain liveness for these keys during the 2 last epochs, it may
+	// exist a doppelganger somewhere.
+	var notRecentStringPubKeys []string
+
+	for _, spk := range stringPubKeys {
+		dph, ok := stringPubKeyToDoppelGangerHelper[spk]
+		if !ok {
+			return nil, errors.New("failed to retrieve doppelganger helper from string public key")
+		}
+
+		if dph.validatorEpoch+2 < currentEpoch {
+			notRecentStringPubKeys = append(notRecentStringPubKeys, spk)
+		}
+	}
+
+	// If all provided keys are recent (aka `notRecentPubKeys` is empty) we return early
+	// as we are unable to effectively determine if a doppelganger is active.
+	if len(notRecentStringPubKeys) == 0 {
+		return buildResponse(stringPubKeys, stringPubKeyToDoppelGangerHelper), nil
+	}
+
+	// Retrieve correspondance between validator pubkey and index
+	stateValidators, err := c.stateValidatorsProvider.GetStateValidators(ctx, notRecentStringPubKeys, nil, nil)
+	if err != nil || stateValidators == nil || stateValidators.Data == nil {
+		return nil, errors.Wrapf(err, "failed to get state validators")
+	}
+
+	validators := stateValidators.Data
+	stringPubKeyToIndex := make(map[string]string, len(validators))
+	indexes := make([]string, len(validators))
+
+	for i, v := range validators {
+		if v == nil {
+			return nil, errors.New("validator is nil")
+		}
+
+		index := v.Index
+
+		if v.Validator == nil {
+			return nil, errors.New("validator JSON is nil")
+		}
+
+		stringPubKeyToIndex[v.Validator.PublicKey] = index
+		indexes[i] = index
+	}
+
+	// Get validators liveness for the the last epoch
+	// We request a state 1 epoch ago. We are guaranteed to have currentEpoch > 2
+	// since we assume that we are not in phase0.
+	previousEpoch := currentEpoch - 1
+
+	indexToPreviousLiveness, err := c.getIndexToLiveness(ctx, previousEpoch, indexes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get map from validator index to liveness for previous epoch (%d)", previousEpoch)
+	}
+
+	// Get validators liveness for the current epoch
+	indexToCurrentLiveness, err := c.getIndexToLiveness(ctx, currentEpoch, indexes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get map from validator index to liveness for current epoch (%d)", currentEpoch)
+	}
+
+	// Set `DuplicateExists` to `true` if needed
+	for _, nrspk := range notRecentStringPubKeys {
```

Comment:
Not a big deal, but inside the loop it does not really matter that the key is not recent, hence a variable name of `pk` would suffice.